pkg/scheduler/queue/queue_test.go

```diff
@@ -240,3 +240,43 @@ func TestRequestQueue_GetNextRequestForQuerier_ShouldReturnImmediatelyIfQuerierI
 	_, _, err := queue.GetNextRequestForQuerier(context.Background(), FirstUser(), querierID)
 	require.EqualError(t, err, "querier has informed the scheduler it is shutting down")
 }
+
+func TestRequestQueue_tryDispatchRequestToQuerier_ShouldReEnqueueAfterFailedSendToQuerier(t *testing.T) {
```

Comment:
I don't love that this test reaches into the internals of the different types here. Could we exercise the same scenario by changing the order of operations? I'm imagining something along the lines of:

1. Call `GetNextRequestForQuerier` with active (not cancelled) context
2. Wait a bit to make sure querier is picked up by dispatcher loop and added to list of waiting queriers
3. Cancel the context
4. Enqueue a request
5. (code without fix would panic here)
6. Call `GetNextRequestForQuerier` again with new context
7. Verify we get the request back

Reply:
this does not work as it hits a different error case about the querier informing it has shutdown. You have to somehow cancel inside `tryDispatchRequestToQuerier`, after `broker.dequeueRequestForQuerier` and before `call.send(reqForQuerier)`.

The only way to do that with the current code structure is to just bypass the querier loop all together, reproduce the steps of `tryDispatchRequestToQuerier` and make the cancel happen in between the two crucial steps.

Reply:
Right, makes sense, thanks.